tests/framework/e2e/etcdctl.go

```diff
@@ -545,9 +554,55 @@ func (ctl *EtcdctlV3) spawnJsonCmd(output interface{}, args ...string) error {
 	if err != nil {
 		return err
 	}
+	defer cmd.Close()
 	line, err := cmd.Expect("header")
 	if err != nil {
 		return err
 	}
 	return json.Unmarshal([]byte(line), output)
 }
+
+func (ctl *EtcdctlV3) Watch(ctx context.Context, key string, opts config.WatchOptions) clientv3.WatchChan {
+	args := ctl.cmdArgs()
+	args = append(args, "watch", key)
+	if opts.RangeEnd != "" {
+		args = append(args, opts.RangeEnd)
+	}
+	args = append(args, "-w", "json")
+	if opts.Prefix {
+		args = append(args, "--prefix")
+	}
+	if opts.Revision != 0 {
+		args = append(args, "--rev", fmt.Sprint(opts.Revision))
+	}
+	proc, err := SpawnCmd(args, nil)
+	if err != nil {
+		return nil
+	}
+
+	ch := make(chan clientv3.WatchResponse)
+	go func() {
```

Comment:
Have you considered to wait for the goroutine to exit when the test case finish?

Reply:
hi, @ahrtr 
 the goroutine will exit by context cancel when the test finishes, see: https://github.com/etcd-io/etcd/pull/14345/files#diff-70546f01c62cf8d453ef453d51b64bbdc664e39613453b572120148a77600a08R76

Reply:
It's possible (although with low possibility) that the test case finish before the goroutine exit, eventually the test might report goroutine leak errors. Our pipeline fails due to this kind of error from time to time. So for safety, we need to make sure the goroutine exits before the the test case returns/exits.

Reply:
I don't think this is a problem. We should be good as long as we propagate signal to goroutine that causes it to exit. It's up to leak detection to wait for this signal to propagate. If you read the etcd test goroutine leak implementation, it gives test up to 5 seconds to cleanup it's goroutines and explicitly gives them time to execute by calling `runtime.Gosched()`.

Reply:
I think we could merge it first. If there is a real problem, I will fix it. What do you think? @serathius @ahrtr

Reply:
I think @ahrtr is worried that the problem will not just imminently show up after merging. Leak goroutine issues like this one are very subtle (for example I cannot reproduce this issue locally at all), so I understand his position.

I still think my argument holds and this should not be a problem, but I would prefer to wait for confirmation from @ahrtr

Reply:
> If you read the etcd test goroutine leak implementation, it gives test up to 5 seconds to cleanup its goroutines and uses `runtime.Gosched()` to give goroutine time to execute and exit.

The goroutine leak checker indeed waits up to [1 second](https://github.com/etcd-io/etcd/blob/e363480089783aff0d1c26586f703d31223cef5d/client/pkg/testutil/leak.go#L117) in the `afterTest` check. 

This is just a mitigation solution instead of best practice to me. We should always make sure gracefully shutdown of all threads/goroutines in almost all cases, including production code and test code, unless the cases that we intentionally create.

Previously we also saw a situation that a test case (in test code) or the main goroutine (in production code) already finishes/exits, but a goroutine it creates still prints log, and it isn't allowed by golang. Of course, I do not see such issue in this PR.

I just searched the repo, there are lots of similar cases which create goroutines but do not gracefully shutdown them. It's OK to merge this PR for now. But I'd like to revisit all such cases afterwards in separate PR(s).  WDYT? @serathius @spzala

Reply:
I agree the is a problem worth looking into.